### ISSUE TITLE
Remote Spring Boot devtools unable to restart application if a dependency has autowire candidates that the main application uses 

### ISSUE THREAD

**Issue by duncan-at-hiveit** [spring-projects/spring-boot#8321]
Hi there,

We have a situation where we have a multi module Gradle project that has shared Spring components, but hot restart fails due to classloader issues.  This is with Spring Boot 1.5.1.RELEASE.

FYI I've put together an example of the error we're having [here](https://github.com/duncan-at-hiveit/spring-boot-devtools-dependency-issue)

Within our project we have some shared `@Services` that are used by several web applications.  These shared services live in their own module within the main project structure and have a bunch of `@Service` classes within it.

A web application will be in a separate module within the main project structure and will have a "compile project" dependency on the shared services module.  It then has `@Controller` controllers that use the `@Service` services from the dependency.  The web application is build with `bootRepackage` and run as a standalone jar, with the shared services module as a jar within `BOOT-INF/lib`.

The problem occurs when updating anything in the shared services module that causes a hot reload.  If that happens, the remote hot reload process (using `RemoteSpringApplication`) tells the `RestartServer` to do a reload of the classes _only within the shared services jar._  The newly reloaded `@Services` are now available in the `RestartClassLoader`.  When Spring fires up to do the hot restart process, it is unable to wire up the `@Controllers` with the `@Services`, because the `@Controllers` are in the `LauchedURLClassLoader` but the `@Services` are now in the `RestartClassLoader`, and so when it's trying to wire up a controller with a particular `@Service` class it is unable to find it because it's no longer in the same classloader as the `@Controller`.

It seems like it would be good if there was some flag to say to restart everything that can be reloadable at the same time, so that you don't end up having some potentially reloadable stuff in one loader and other potentially reloadable stuff in the the loader. 

Many thanks for your time and help,
Duncan

**Comment by snicoll** [comment]
What you are describing is explicitly [described in the documentation](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-devtools-customizing-classload) so please review that and update your project accordingly.

**Comment by duncan-at-hiveit** [comment]
Hi Stephane,

Thanks for the update.  I do have the entry:

restart.include.hotdeployjars=/hotreloaddependency\.jar

in src/main/resources/META-INF/spring-devtools.properties, which matches the name of the jar that the dependency produces.  Have I miscofigured this somehow?

Many thanks,
Duncan

**Comment by duncan-at-hiveit** [comment]
Hi there Stephane,

Thanks for reopening.  I suspect that the issue may be related to the remote nature of using devtools.  For instance, when running bootRun directly, I see logging like:

15:41:50.761 [QUIET] [system.out] 15:41:50.760 [main] DEBUG org.springframework.boot.devtools.settings.DevToolsSettings - Included patterns for restart : [/hotreloaddependency.jar]
15:41:50.762 [QUIET] [system.out] 15:41:50.761 [main] DEBUG org.springframework.boot.devtools.settings.DevToolsSettings - Excluded patterns for restart : [/spring-boot-starter/target/classes/, /spring-boot-au

When firing up the jar remotely I see no such logging.  Obviously devtools is there because we're able to break it with redeploys, but I wonder if DevToolsSettings is never invoked?

Many thanks,
Duncan

**Comment by snicoll** [comment]
I have the same analysis and I can't see how this would ever work remotely actually. But looking at the documentation, it isn't clear either it would not be supported so I've reopened it to clarify that.

**Comment by duncan-at-hiveit** [comment]
Thanks!

**Comment by duncan-at-hiveit** [comment]
FYI yesterday I raised a similar question in Stack Overflow regarding the documentation of local versus remote capabilities, so if you're able to provide clarification here I will update my post at:

http://stackoverflow.com/questions/42276108/limitations-of-spring-boot-devtools-with-a-remote-application

Thanks,
Duncan

**Comment by philwebb** [comment]
Just digging into the first issue from the sample. I think we do have some bugs with restart logic, probably because the multi-module remote use-case hasn't been considered that well.

The first thing of note is that the `META-INF/spring-devtools.properties` file isn't considered when working in the remote mode. Even if it were the entry won't fix the issue because it's the main application classes that need to be pushed up (not just the `hotreloaddependency` jar).

If you start the application, the initial load doesn't do anything special. The restart classloader isn't installed.

When you change a file in `hotreloaddependency` a restart is triggered. The restart classloader _only_ contains the `hotreloaddependency` jar so things fall over.

If you restart the experiment but instead first change a file in `hotreloaddependant` first, things work fine. This is because the change in `hotreloaddependant` pushes that jar up, then the second update to `hotreloaddependency` is fine because both are in the same classloader.

**Comment by philwebb** [comment]
It might be a while before we can get a fix in for this one. It's unfortunately going to get a lower priority than other, more critical bugs.

**Comment by duncan-at-hiveit** [comment]
Hey that's brilliant, thanks!

**Comment by KramKroc** [comment]
Encountered this in the last few days. At some point I was able to make the changes and see them load remotely without the nullpointer being triggered but since the weekend it's failing over with each update.

There's a hint of a workaround in what @philwebb mentioned:

> If you restart the experiment but instead first change a file in hotreloaddependant first, things work fine. This is because the change in hotreloaddependant pushes that jar up, then the second update to hotreloaddependency is fine because both are in the same classloader.

and I wondered if that could be clarified. As I said, I had it working in some form last week but I guess whatever process I was doing to do updates has changed and I can't unpick it :(

**Comment by KayanoLiam** [comment]
Reproduced this on current 4.1.x: a first update containing only a dependency module leaves unchanged application beans in the parent classloader. I'm working on including all watched source-directory names in upload metadata, with a regression exercising serialization, restart classloading, and constructor injection across the two modules.

**Comment by wilkinsona** [comment]
Thanks for taking a look, @KayanoLiam, but I'm not sure that'll be a good use of your time. We've not seen much demand for this one, with the last comment being 8 years ago and with only a couple of people apparently affected.

**Comment by wilkinsona** [comment]
We’re cleaning out the issue tracker and closing issues that we’ve not seen much demand to fix. Feel free to comment with additional justifications if you feel that this one should not have been closed.
